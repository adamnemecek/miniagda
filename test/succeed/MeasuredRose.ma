-- 2010-07-27

data List (+ A : Set) : Set
{ nil  : List A
; cons : A -> List A -> List A
}

fun mapList : [A : Set] -> [B : Set] -> (A -> B) -> List A -> List B
{ mapList A B f (nil .A) = nil B
; mapList A B f (cons .A a as) = cons B (f a) (mapList A B f as)
}

-- sized Roses

sized data Rose (+ A : Set) : Size -> Set
{ rose : [i : Size] -> A -> List (Rose A i) -> Rose A ($ i) 
}

fun mapRose : [A : Set] -> [B : Set] -> (A -> B) -> 
              [i : Size] -> |i| -> Rose A i -> Rose B i
{ mapRose A B f i (rose .A (i > j) a rs) = 
    rose B j (f a) (mapList (Rose A j) (Rose B j) (mapRose A B f j) rs)
}