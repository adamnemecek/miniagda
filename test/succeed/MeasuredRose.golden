--- opening "MeasuredRose.ma" ---
--- scope checking ---
--- type checking ---
type  List : ++(A : Set) -> Set
term  List.nil : .[A : Set] -> < List.nil : List A >
term  List.cons : .[A : Set] -> ^(y0 : A) -> ^(y1 : List A) -> < List.cons y0 y1 : List A >
term  mapList : .[A : Set] -> .[B : Set] -> (A -> B) -> List A -> List B
{ mapList [A] [B] f List.nil = List.nil
; mapList [A] [B] f (List.cons a as) = List.cons (f a) (mapList [A] [B] f as)
}
type  Rose : ++(A : Set) -> + Size -> Set
term  Rose.rose : .[A : Set] -> .[s!ze : Size] -> .[i < s!ze] -> ^ A -> ^ List (Rose A i) -> Rose A s!ze
term  Rose.rose : .[A : Set] -> .[i : Size] -> ^(y1 : A) -> ^(y2 : List (Rose A i)) -> < Rose.rose i y1 y2 : Rose A $i >
term  mapRose : .[A : Set] -> .[B : Set] -> (A -> B) -> .[i : Size] -> Rose A i -> Rose B i
{ mapRose [A] [B] f [i] (Rose.rose [j < i] a rs) = Rose.rose [j] (f a) (mapList [Rose A j] [Rose B j] (mapRose [A] [B] f [j]) rs)
}
term  mapRose' : .[A : Set] -> .[B : Set] -> (A -> B) -> .[i : Size] -> Rose A i -> Rose B i
{ mapRose' [A] [B] f [i] (Rose.rose [j < i] a rs) = Rose.rose [j] (f a) (mapList [Rose A j] [Rose B j] (mapRose' [A] [B] f [j]) rs)
}
--- evaluating ---
--- closing "MeasuredRose.ma" ---
